File: src/main/java/fr/aven/bot/music/PlayerManager.java
package fr.aven.bot.music;

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import com.wrapper.spotify.model_objects.specification.Paging;
import com.wrapper.spotify.model_objects.specification.PlaylistTrack;
import com.wrapper.spotify.model_objects.specification.Track;
import fr.aven.bot.Main;
import fr.aven.bot.util.MessageTask;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.MessageBuilder;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.events.message.guild.GenericGuildMessageEvent;

import javax.annotation.Nullable;
import java.awt.*;
import java.util.*;
import java.util.List;

public class PlayerManager
{
    private static PlayerManager INSTANCE;
    private final AudioPlayerManager playerManager;
    private final Map<Long, GuildMusicManager> musicManagers;

    private PlayerManager()
    {
        this.musicManagers = new HashMap<>();

        this.playerManager = new DefaultAudioPlayerManager();
        AudioSourceManagers.registerLocalSource(playerManager);
        AudioSourceManagers.registerRemoteSources(playerManager);
    }

    public boolean checkNullForEvent(Guild guild)
    {
        return musicManagers.get(guild.getIdLong()) == null;
    }

    /*public void destroyGuildMusicManager(Guild guild)
    {
        musicManagers.remove(guild.getIdLong());
    }*/

    public synchronized GuildMusicManager getGuildMusicManager(Guild guild, TextChannel channel)
    {
        long guildId = guild.getIdLong();
        GuildMusicManager musicManager = musicManagers.get(guildId);

        if (musicManager == null)
        {
            musicManager = new GuildMusicManager(playerManager, guild, channel);
            musicManagers.put(guildId, musicManager);
        }

        guild.getAudioManager().setSendingHandler(musicManager.getSendHandler());

        return musicManager;
    }

    public void loadAndPlaySpotifyPlaylist(Message message, Paging<PlaylistTrack> playlistTracks)
    {
        message.getChannel().sendMessage("Playlist added ! Please wait...").queue();

        List<AudioTrack> audioTracks = new ArrayList<>();
        GuildMusicManager musicManager = getGuildMusicManager(message.getGuild(), message.getTextChannel());


        for (PlaylistTrack playlistTrack : playlistTracks.getItems())
        {
            if (playlistTrack.getIsLocal()) continue;
            Track track = (Track) playlistTrack.getTrack();

            String search = "ytsearch:"+track.getName()+" "+track.getArtists()[0].getName();

            playerManager.setFrameBufferDuration(5000);
            playerManager.loadItemOrdered(musicManager, search, new AudioLoadResultHandler()
            {
                @Override
                public void trackLoaded(AudioTrack track) {
                }

                @Override
                public void playlistLoaded(AudioPlaylist playlist) {
                    if (playlist.isSearchResult())
                        audioTracks.add(playlist.getTracks().get(0));

                }

                @Override
                public void noMatches() {
                }

                @Override
                public void loadFailed(FriendlyException exception) {
                }
            });

        }

        AudioTrack firstTrack = audioTracks.get(0);
        musicManager.scheduler.usersRequest.put(firstTrack, message.getAuthor().getIdLong());

        for (int i = 1; i < audioTracks.size(); i++)
            musicManager.scheduler.usersRequest.put(audioTracks.get(i), message.getAuthor().getIdLong());

        play(musicManager, firstTrack, message.getTextChannel());
        audioTracks.forEach(musicManager.scheduler::queue);
    }

    public void loadAndPlaySpotifyTrack(Message message, Track track)
    {
        String search = "ytsearch:"+ track.getName()+" "+ track.getArtists()[0].getName();

        loadAndPlay(message,search, true);
    }

    public void loadAndPlay(Message message, String trackUrl, boolean spotify)
    {
        GuildMusicManager musicManager = getGuildMusicManager(message.getGuild(), message.getTextChannel());

        playerManager.setFrameBufferDuration(5000);
        playerManager.loadItemOrdered(musicManager, trackUrl, new AudioLoadResultHandler()
        {
            @Override
            public void trackLoaded(AudioTrack track)
            {
                String title;
                String author;
                EmbedBuilder builder = new EmbedBuilder();
                builder.setThumbnail("https://i.ytimg.com/vi/" + track.getInfo().identifier + "/maxresdefault.jpg");
                title = track.getInfo().title;
                author = track.getInfo().author;

                message.getChannel().sendMessage(builder.setAuthor(Main.getDatabase().getTextFor("music.add", message.getGuild()), track.getInfo().uri, message.getJDA().getSelfUser().getAvatarUrl())
                        .addField("❱ "+Main.getDatabase().getTextFor("music.author", message.getGuild())+" : "+author, "❱ "+title, false)
                        .setColor(new Color(0, 255, 151))
                        .setFooter("AvenBot by Aven#1000").build()).queue(msg -> {
                            new Timer().schedule(new MessageTask(msg), 10000);
                });

                musicManager.scheduler.usersRequest.put(track, message.getAuthor().getIdLong());
                play(musicManager, track, message.getTextChannel());
            }

            @Override
            public void playlistLoaded(AudioPlaylist playlist) {
                if (playlist.isSearchResult()) {
                    if (spotify)
                    {
                        AudioTrack audioTrack = playlist.getTracks().get(0);


                            String title;
                            String author;
                            EmbedBuilder builder = new EmbedBuilder();
                            builder.setThumbnail("https://i.ytimg.com/vi/" + audioTrack.getInfo().identifier + "/maxresdefault.jpg");
                            title = audioTrack.getInfo().title;
                            author = audioTrack.getInfo().author;

                            message.getChannel().sendMessage(builder.setAuthor(Main.getDatabase().getTextFor("music.add", message.getGuild()) + " (Spotify)", audioTrack.getInfo().uri, message.getJDA().getSelfUser().getAvatarUrl())
                                    .addField("❱ " + Main.getDatabase().getTextFor("music.author", message.getGuild()) + " : " + author, "❱ " + title, false)
                                    .setColor(new Color(0, 255, 151))
                                    .setFooter("AvenBot by Aven#1000").build()).queue(msg -> {
                                new Timer().schedule(new MessageTask(msg), 10000);
                            });

                        musicManager.scheduler.usersRequest.put(audioTrack, message.getAuthor().getIdLong());
                        play(musicManager, audioTrack, message.getTextChannel());
                    } else {

                        EmbedBuilder builder = new EmbedBuilder();
                        builder.setAuthor(Main.getDatabase().getTextFor("music.searchTitle", message.getGuild()), "https://justaven.com", message.getAuthor().getAvatarUrl());
                        builder.setColor(message.getMember().getColor());
                        builder.setFooter(Main.getDatabase().getTextFor("music.searchFooter", message.getGuild()), message.getJDA().getSelfUser().getAvatarUrl());

                        for (int i = 0; i < playlist.getTracks().size() && i < 5; i++) {
                            AudioTrack track = playlist.getTracks().get(i);
                            AudioTrackInfo info = track.getInfo();
                            int nbTrack = i;
                            nbTrack++;
                            builder.appendDescription("\n`" + nbTrack + "`: **" + info.title + "** | " + Main.getDatabase().getTextFor("music.author", message.getGuild()) + " : " + info.author);

                            musicManager.scheduler.search.put(nbTrack, track);
                        }

                        message.getTextChannel().sendMessage(builder.build()).queue(msg -> {
                            msg.addReaction("❌").queue();
                            musicManager.scheduler.lastMessageSearch = msg;
                        });
                    }
                } else {
                    AudioTrack firstTrack = playlist.getSelectedTrack();

                    if (firstTrack == null) {
                        firstTrack = playlist.getTracks().remove(0);
                    }

                    //message.getChannel().sendMessageFormat(Main.getDatabase().getTextFor("music.playlistAdd", message.getGuild()), firstTrack.getInfo().title, playlist.getName()).queue();

                    EmbedBuilder builder = new EmbedBuilder();
                    builder.setAuthor(Main.getDatabase().getTextFor("playlist.title", message.getGuild()));
                    builder.addField("❱ "+playlist.getName() + " (" + playlist.getTracks().size() + " tracks)",
                            "❱ "+ Main.getDatabase().getTextFor("playlist.firstTrack", message.getGuild()) + " : " + playlist.getTracks().get(0).getInfo().title,
                            false).setFooter("AvenBot by Aven#1000").setColor(new Color(255, 127, 0));

                    message.getChannel().sendMessage(builder.build()).queue(msg -> {
                        new Timer().schedule(new MessageTask(msg), 10000);
                    });

                    musicManager.scheduler.usersRequest.put(firstTrack, message.getAuthor().getIdLong());
                    for (AudioTrack track : playlist.getTracks())
                        musicManager.scheduler.usersRequest.put(track, message.getAuthor().getIdLong());

                    play(musicManager, firstTrack, message.getTextChannel());
                    playlist.getTracks().forEach(musicManager.scheduler::queue);

                }

            }
            @Override
            public void noMatches()
            {
                message.getChannel().sendMessage(Main.getDatabase().getTextFor("music.notFound", message.getGuild())).queue();
            }

            @Override
            public void loadFailed(FriendlyException exception)
            {
                exception.printStackTrace();
                message.getChannel().sendMessage(Main.getDatabase().getTextFor("music.couldntPlay", message.getGuild()) + exception.getMessage()).queue();
            }
        });

    }

    private void play(GuildMusicManager musicManager, AudioTrack track, TextChannel channel)
    {
        if (musicManager.scheduler.getQueue().isEmpty()) musicManager.scheduler.channel = channel;
        musicManager.scheduler.queue(track);
    }

    public static synchronized PlayerManager getInstance()
    {
        if (INSTANCE == null)
        {
            INSTANCE = new PlayerManager();
        }

        return INSTANCE;
    }
}

File: src/main/kotlin/fr/aven/bot/music/PlayerManager.kt
package fr.aven.bot.music

import api.deezer.objects.Track
import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers
import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeAudioSourceManager
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.lava.extensions.youtuberotator.YoutubeIpRotatorSetup
import com.sedmelluq.lava.extensions.youtuberotator.planner.NanoIpRoutePlanner
import com.sedmelluq.lava.extensions.youtuberotator.tools.ip.Ipv6Block
import dev.minn.jda.ktx.Embed
import dev.minn.jda.ktx.SLF4J
import fr.aven.bot.core.Config
import fr.aven.bot.util.Language
import fr.aven.bot.util.lang.LangKey
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.TextChannel
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent
import net.dv8tion.jda.api.interactions.commands.SlashCommandInteraction
import java.awt.Color
import java.time.Instant

class PlayerManager(private val config: Config)
{
    private val logger by SLF4J
    private var playerManager: AudioPlayerManager = DefaultAudioPlayerManager()
    private var musicManagers = mutableMapOf<Long, GuildMusicManager>()

    private val youtubeAudioSourceManager = YoutubeAudioSourceManager(true)

    init {
        logger.info("Init PlayerManager...")

        logger.info("Init YouTube IP Rotator...")
        val ipv6 = config.ipv6_block
        if (ipv6 == "none") logger.warn("Skipped IP Rotator")
        else YoutubeIpRotatorSetup(NanoIpRoutePlanner(listOf(Ipv6Block(ipv6)), true)).forSource(youtubeAudioSourceManager).setup()

        this.playerManager.registerSourceManager(youtubeAudioSourceManager)
        logger.info("Initialized!")

        AudioSourceManagers.registerLocalSource(playerManager)
        AudioSourceManagers.registerRemoteSources(playerManager)
    }

    fun guildMusicManager(event: ButtonInteractionEvent): GuildMusicManager = guildMusicManager(event.guild!!, event.textChannel)
    fun guildMusicManager(event: SelectMenuInteractionEvent): GuildMusicManager = guildMusicManager(event.guild!!, event.textChannel)
    fun guildMusicManager(interaction: SlashCommandInteraction): GuildMusicManager = guildMusicManager(interaction.guild!!, interaction.textChannel)
    private fun guildMusicManager(guild: Guild, channel: TextChannel): GuildMusicManager
    {
        musicManagers.putIfAbsent(guild.idLong, GuildMusicManager(playerManager, guild, channel))
        guild.audioManager.sendingHandler = musicManagers[guild.idLong]!!.sendHandler()
        return musicManagers[guild.idLong]!!
    }

    fun loadAndPlayDeezerTrack(interaction: SlashCommandInteraction, track: Track) = loadAndPlay(interaction, "ytsearch:${track.title} ${track.artist.name}")

    fun loadAndPlay(interaction: SlashCommandInteraction, trackUrl: String)
    {
        val guild = interaction.guild!!
        val musicManager = guildMusicManager(interaction)
        playerManager.frameBufferDuration = 5000
        playerManager.loadItemOrdered(musicManager, trackUrl, object : AudioLoadResultHandler
        {
            override fun trackLoaded(track: AudioTrack) {
                interaction.replyEmbeds(musicManager.embedConfirm(track)).setEphemeral(true).queue()

                play(musicManager, track, interaction)
            }

            override fun playlistLoaded(playlist: AudioPlaylist) {
                if (playlist.isSearchResult) musicManager.searchResult(playlist, interaction)
                else {
                    var firstTrack: AudioTrack? = playlist.selectedTrack;
                    if (firstTrack == null)
                        firstTrack = playlist.tracks.removeFirst()

                    interaction.replyEmbeds(Embed {
                        author {
                            name = musicManager.language.getString(LangKey.keyBuilder(this, "playlistLoaded", "playlist.title"), "Playlist added")
                        }

                        field {
                            name = "❱ ${playlist.name} (${playlist.tracks.size} tracks)"
                            value = "❱ ${musicManager.language.getString(LangKey.keyBuilder(this, "playlistLoaded", "playlist.firstTrack"), "First track")} : ${firstTrack!!.info.title}"
                            inline = false
                        }

                        footer {
                            name = "AvenBot by Aven#1000"
                        }

                        color = Color(255, 127, 0).rgb
                    }).setEphemeral(true).queue()

                    play(musicManager, firstTrack!!, interaction)
                    playlist.tracks.forEach { musicManager.scheduler.queue(it, interaction.textChannel, interaction.member!!) }
                }
            }

            override fun noMatches() {
                interaction.reply(musicManager.language.getString(LangKey.keyBuilder(this, "noMatches", "music.notFound"), "Nothing found, please retry (link recommended).")).queue()
            }

            override fun loadFailed(exception: FriendlyException?) {
                interaction.reply(musicManager.language.getString(LangKey.keyBuilder(this, "noMatches","music.couldntPlay"), "Could not play.")).queue()
            }

        })
    }

    fun play(musicManager: GuildMusicManager, track: AudioTrack, interaction: SlashCommandInteraction) = musicManager.scheduler.queue(track, interaction.textChannel, interaction.member!!)
}